### PR TITLE
Add circuit for RETURNDATASIZE opcode

### PR DESCRIPTION
**Pull request by xiaodino**

### CODE REVIEW

File: bus-mapping/src/evm/opcodes/returndatasize.rs
Comment (CPerezz) [inline bus-mapping/src/evm/opcodes/returndatasize.rs:60]: I think this could be a bit more elaborated. 
We should create a sub-context with either: [CALL](https://www.evm.codes/#F1), [CALLCODE](https://www.evm.codes/#F2), [DELEGATECALL](https://www.evm.codes/#F4) or [STATICCALL](https://www.evm.codes/#FA) and then get a `RETURNDATASIZE` from there. I'd also try to play with really big values and values like 0 or similar. So that we test all the possible edge-cases.

File: bus-mapping/src/evm/opcodes/returndatasize.rs
Comment (CPerezz) [inline bus-mapping/src/evm/opcodes/returndatasize.rs:67]: There's no call data even in the tx here. 
So as said avobe this could be a bit more curated.

File: zkevm-circuits/src/evm_circuit/execution.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution.rs:455]: I'd remove these two empty lines added.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:81]: It's fine if we want to panic here. But I'd use `expect` and add a small description of the panic.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:99]: As said before, I'd try to test for edge cases as really big returndata size numbers or 0 or things like this.
Otherwise, the testing is not as useful as it could be.

**Review by CPerezz — CHANGES_REQUESTED** [review]
Great work! And thanks for submitting a PR for it. 

I have some things that should probably be addressed. Specifically with the testing cases applied here.

Aside from that the rest looks good! :smile:

File: bus-mapping/src/evm/opcodes/returndatasize.rs
Comment (xiaodino) [reply bus-mapping/src/evm/opcodes/returndatasize.rs:60]: Thanks. I'm working on it.

Comment (xiaodino) [reply bus-mapping/src/evm/opcodes/returndatasize.rs]: Updated the tests. Create a sub-context with ```CALL``` and then get a ```RETURNDATASIZE``` from there.

File: zkevm-circuits/src/evm_circuit/execution.rs
Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution.rs]: Removed.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution/returndatasize.rs]: Updated.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution/returndatasize.rs]: Updated tests.

**Review by xiaodino — COMMENTED** [review]
@CPerezz  Thanks for the feedbacks. They have been addressed.

**Review by CPerezz — APPROVED** [review]
LGTM!! Thanks for your work on the comments @xiaodino 

Requesting one more review before merging.
Maybe @roynalnaruto can check this?

**Comment by roynalnaruto** [comment]
I'll review this PR before end of today.

**Comment by CPerezz** [comment]
@xiaodino please can you fix the issues with `rustfmt`?

**Comment by xiaodino** [comment]
> 

The issues with rustfmt have been fixed.

**Comment by ChihChengLiang** [comment]
@roynalnaruto is the review of this pr blocked?

**Comment by roynalnaruto** [comment]
> @roynalnaruto is the review of this pr blocked?

No, I was just busy working on the exponentiation feature and completely forgot about this. Will review by tomorrow end :)

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:23]:
```suggestion
    return_data_size: RandomLinearCombination<F, N_BYTES_U64>,
```

Geth uses `uint64` for the length of return data:
https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L318

I think we should use `N_BYTES_U64`

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:77]:
```suggestion
                return_data_size.to_le_bytes()[..N_BYTES_U64]
```

And the corresponding change here as well

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:185]: Not a biggie, but I think you can simplify this right? Or am I missing something?

```suggestion
    assert_eq!(
        run_test_circuits(
            TestContext::<2, 1>::simple_ctx_with_bytecode(code).unwrap(),
            None              
        ),                                                                                                                 
        Ok(())
    );
```

**Review by roynalnaruto — COMMENTED** [review]
Overall looks good! A few comments, a few suggestions (especially the change from `N_BYTES_MEMORY_ADDRESS` to `N_BYTES_U64`) and a few questions.

Good work! And thanks for contributing :)

File: bus-mapping/src/evm/opcodes/return.rs
Comment (xiaodino) [reply bus-mapping/src/evm/opcodes/return.rs]: Good catch. Updated.

Thanks.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution/returndatasize.rs]: Make sense. Updated.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution/returndatasize.rs]: Updated.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution/returndatasize.rs]: Updated. Thanks.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/returndatasize.rs]: I think that it should not be really necessary. As something bigger than U64 should imply that the memory expansion cost would be inassumible. 

And, so, we save some RLC saves for example.

File: zkevm-circuits/src/evm_circuit/execution/returndatasize.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/returndatasize.rs]: ditto

**Comment by CPerezz** [comment]
@roynalnaruto I actually don't really agree with the U256 instead of U64 used for the value RLC.

See: https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/703#discussion_r968742152

**Comment by roynalnaruto** [comment]
> @roynalnaruto I actually don't really agree with the U256 instead of U64 used for the value RLC.
> 
> See: [#703 (comment)](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/703#discussion_r968742152)

I mentioned that since geth uses `uint64`, so I thought we should stick to 8 bytes (`N_BYTES_U64`). `N_BYTES_MEMORY_ADDRESS` was 5 bytes.

**Comment by Brechtpd** [comment]
I think in most opcodes we do use `N_BYTES_MEMORY_WORD_SIZE` for storing a similar value (memory length/memory offsets in `MemoryAddress` types). Though a small difference here is that this opcode on its own does not not need any memory expansion (which is where the gas cost would overflow a 64bit value for these 5-byte sizes/offsets already) because it's a simple getter. So in that aspect `N_BYTES_U64` could make sense instead, but of course the `RETURN` will already have done the memory expansion so it is as impossible to have bigger values than `N_BYTES_MEMORY_WORD_SIZE` as in all other uses cases where that is used I believe.

So good points to make for both I feel, but practically `N_BYTES_MEMORY_WORD_SIZE` should be as okay as its use in other opcodes. On the other hand, using `N_BYTES_U64` won't make that much difference.

File: bus-mapping/src/evm/opcodes/return.rs
Comment (han0110) [inline bus-mapping/src/evm/opcodes/return.rs:66]: After rebase on `main`, these should be handled in `handle_restore_context` and could be removed.

**Review by han0110 — COMMENTED** [review]
LGTM! After rebase we could merge this.

File: bus-mapping/src/evm/opcodes/return.rs
Comment (xiaodino) [reply bus-mapping/src/evm/opcodes/return.rs:66]: Is this PR: https://github.com/xiaodino/zkevm-circuits/commit/809895b75630c6ee42eadfba85e8274389a20c5b the one that handle `state.call_context_write`?

Comment (xiaodino) [reply bus-mapping/src/evm/opcodes/return.rs]: Updated. Please take a look

**Review by han0110 — APPROVED** [review]
LGTM! Once the CI passes we can merge this.

**Comment by xiaodino** [comment]
> LGTM! Once the CI passes we can merge this.

CI passed!

**Pull request merged**
